lib/csv/decoder.ex

```diff
@@ -61,120 +61,102 @@ defmodule CSV.Decoder do
       [%{:x => \"a\", :y => \"b\"}, %{:x => \"c\", :y => \"d\"}]
   """
 
-  def decode(stream, options \\ []) do
-    { headers, stream } = options
-                          |> Keyword.get(:headers, false)
-                          |> get_headers(stream, options)
-
-    row_length = case headers do
-      false -> stream |> get_row_length(options)
-      _ -> headers |> Enum.count
-    end
-
-    num_workers = options
-                  |> Keyword.get(:num_workers, options
-                  |> Keyword.get(:num_pipes, Defaults.num_workers))
-
-    multiline_escape = options
-                        |> Keyword.get(:multiline_escape, true)
+  def decode!(stream, options \\ []) do
+    decode(stream, options)
+    |> handle_errors!
+  end
 
+  def decode(stream, options \\ []) do
+    { stream, options |> options_with_defaults }
+    |> prepare_headers
+    |> prepare_row_length
+    |> decode_stream
+  end
 
+  defp decode_stream({ stream, %{ multiline_escape: multiline_escape } = options }) do
     stream
     |> aggregate(multiline_escape)
     |> Stream.with_index
-    |> ParallelStream.map(fn { line, index } ->
-      { line, index }
-      |> Lexer.lex(options)
-      |> lex_to_parse
-      |> Parser.parse(options)
-      |> check_row_length(row_length)
-      |> build_row(headers)
-    end, options |> Keyword.merge(num_workers: num_workers))
-    |> handle_errors
-  end
-  defp aggregate(stream, true) do
-    stream |> LineAggregator.aggregate
-  end
-  defp aggregate(stream, false) do
-    stream
+    |> ParallelStream.map(&(decode_row(&1, options)),
+                          options |> Enum.into([]))
   end
 
-  defp lex_to_parse({ :ok, tokens, index }) do
-    { tokens, index }
-  end
-  defp lex_to_parse(result) do
-    result
+  defp decode_row({ nil, 0 }, %{ row_length: false }), do: { :ok, [] }
+  defp decode_row({ line, index }, %{ headers: headers, row_length: row_length } = options) do
+    with { :ok, parsed, _ } <- parse_row({ line, index }, options |> Enum.into([])),
```

Comment:
I agree that the `2.0` release might be a good time to change Elixir version compatibility, I'd like to do some research on the topic before making a decision though. Let's leave this as is, I'll get back to you if it needs to be changed.